jwql/website/apps/jwql/forms.py

```diff
@@ -137,26 +126,89 @@ class AnomalyQueryForm(BaseForm):
+
+    niriss_detector_list = []
+    for detector in DETECTOR_PER_INSTRUMENT['niriss']:
+        detector = query_format(detector)
+        niriss_detector_list.append([detector, detector])
+
+    nircam_detector_list = []
+    for detector in DETECTOR_PER_INSTRUMENT['nircam']:
+        detector = query_format(detector)
+        nircam_detector_list.append([detector, detector])
+
+    nirspec_detector_list = []
+    for detector in DETECTOR_PER_INSTRUMENT['nirspec']:
+        detector = query_format(detector)
+        nirspec_detector_list.append([detector, detector])
+
+    # Generate dynamic lists of read patterns to use in forms
+    miri_readpatt_list = []
+    for readpatt in READPATT_PER_INSTRUMENT['miri']:
+        readpatt = query_format(readpatt)
+        miri_readpatt_list.append([readpatt, readpatt])
+
+    niriss_readpatt_list = []
+    for readpatt in READPATT_PER_INSTRUMENT['niriss']:
+        readpatt = query_format(readpatt)
+        niriss_readpatt_list.append([readpatt, readpatt])
+
+    nircam_readpatt_list = []
+    for readpatt in READPATT_PER_INSTRUMENT['nircam']:
+        readpatt = query_format(readpatt)
+        nircam_readpatt_list.append([readpatt, readpatt])
+
+    nirspec_readpatt_list = []
+    for readpatt in READPATT_PER_INSTRUMENT['nirspec']:
+        readpatt = query_format(readpatt)
+        nirspec_readpatt_list.append([readpatt, readpatt])
+
+    # Generate dynamic lists of exposure types to use in forms
+    miri_exptype_list = []
+    for exptype in EXP_TYPE_PER_INSTRUMENT['miri']:
+        exptype = query_format(exptype)
+        miri_exptype_list.append([exptype, exptype])
+
+    niriss_exptype_list = []
+    for exptype in EXP_TYPE_PER_INSTRUMENT['niriss']:
+        exptype = query_format(exptype)
+        niriss_exptype_list.append([exptype, exptype])
+
+    nircam_exptype_list = []
+    for exptype in EXP_TYPE_PER_INSTRUMENT['nircam']:
+        exptype = query_format(exptype)
+        nircam_exptype_list.append([exptype, exptype])
+
+    nirspec_exptype_list = []
+    for exptype in EXP_TYPE_PER_INSTRUMENT['nirspec']:
+        exptype = query_format(exptype)
+        nirspec_exptype_list.append([exptype, exptype])
+
+    # Generate dynamic lists of grating options to use in forms
+    miri_grating_list = []
+    for grating in GRATING_PER_INSTRUMENT['miri']:
+        grating = query_format(grating)
+        miri_grating_list.append([grating, grating])
+
+    niriss_grating_list = []
+    for grating in GRATING_PER_INSTRUMENT['niriss']:
+        grating = query_format(grating)
+        niriss_grating_list.append([grating, grating])
+
+    nircam_grating_list = []
+    for grating in GRATING_PER_INSTRUMENT['nircam']:
+        grating = query_format(grating)
+        nircam_grating_list.append([grating, grating])
+
+    nirspec_grating_list = []
+    for grating in GRATING_PER_INSTRUMENT['nirspec']:
+        grating = query_format(grating)
+        nirspec_grating_list.append([grating, grating])
```

Comment:
I'm thinking that using a dictionary here might allow you to be able to loop over instrument and remove some similar lines of code, something like:

```python
params = {}
for instrument in ['miri', 'niriss', 'nircam', 'nirspec']:
    params[instrument] = {}
    params[instrument]['detector_list'] = []
    for detector in DETECTOR_PER_INSTRUMENT[instrument]:
        detector = query_format(detector)
        params[instrument]['detector_list'].append([detector, detector])
```

Reply:
Absolutely, that's a great idea! I was thinking this was getting a bit excessive, and a dictionary would cut down quite a bit of repetitive code.